### ISSUE TITLE
Drop user-facing `hoist`, use Popover API (keep `hoist` code as internal polyfill)

### ISSUE THREAD

**Issue by LeaVerou** [shoelace-style/webawesome#273]
We currently have 5 components that use `hoist`:

- [ ] [`<wa-color-picker>`](https://webawesome-git-next-font-awesome.vercel.app/docs/components/color-picker) (to pass it to its internal `<wa-dropdown>`)
- [ ] [`<wa-copy-button>`](https://webawesome-git-next-font-awesome.vercel.app/docs/components/copy-button) (to pass it to its internal `<wa-tooltip>`)
- [ ] [`<wa-dropdown>`](https://webawesome-git-next-font-awesome.vercel.app/docs/components/dropdown) (to pass it to its internal `<wa-popup>`)
- [ ] [`<wa-select>`](https://webawesome-git-next-font-awesome.vercel.app/docs/components/select) (to pass it to its internal `<wa-popup>`)
- [ ] [`<wa-tooltip>`](https://webawesome-git-next-font-awesome.vercel.app/docs/components/tooltip) (to pass it to its internal `<wa-popup>`)
- [ ] [`<wa-popup>`](https://webawesome-git-next-font-awesome.vercel.app/docs/components/popup#properties) calls it `strategy` but it's the same thing (`hoist` == `strategy=fixed`)

Presumably because of implementation complexity / perf, this is not the default behavior, which introduces subtle bugs that authors may not even notice (or may not know how to fix). For example, this bug in our own docs, that we had to fix with a JS snippet that just adds `hoist` to all selects:

<img width="730" alt="Image" src="https://github.com/user-attachments/assets/7f7b6538-8779-48f1-9f4f-c6747ad4f0bb" />

At this point the [Popover API is supported everywhere](https://caniuse.com/mdn-api_htmlelement_popover) and also fixes it properly, as it uses a UA-level primitive that regular CSS does not have access to. Are there any blockers from us using just that, or is it just about doing the work?

**Comment by claviska** [comment]
I would love to get these updated!

> Presumably because of implementation complexity, this is not the default behavior

Hoist is a fixed positioning strategy (read: hack) that a lot of people seem to use. I can confirm that the Popover API does indeed solve it, but I remain concerned about browser support so I've only implemented it in experimental components.

**88% supported** per https://caniuse.com/mdn-api_htmlelement_popover

We really need to establish #147 and make sure _our users_ and _their clients_ are OK with it before we ship things we can't easily workaround.

**Comment by claviska** [comment]
Oh, and it's not the default behavior because it's less performant, particularly when scrolling, and the default strategy works 95% of the time (basically unless it's getting clipped by an overflow).

The popup component _does_ have hoist, but it's defined as:

```ts
strategy: 'fixed' | 'absolute'
```

In hindsight, it probably could have been named the same thing, but hoist was always intended to be a last resort, remove-it-as-soon-as-possible API that was easy to use in a pinch.

**Comment by LeaVerou** [comment]
> **88% supported** per [caniuse.com/mdn-api_htmlelement_popover](https://caniuse.com/mdn-api_htmlelement_popover)
> 
> We really need to establish #147 and make sure _our users_ and _their clients_ are OK with it before we ship things we can't easily workaround.

What we should be looking at rn is whether all major engines support it, the exact % doesn't matter because if all three support it, by the time WA ships it will be > 90%. The only features we can't count on are those not supported by one of the engines (and even then, I'd be inclined to still use it if the holdout is FF and we know it's currently being worked on, e.g. View Transitions)

**Comment by claviska** [comment]
I normally agree with you — Shoelace was created as a "bleeding edge" library. As we mature and have paying users, I'm more sensitive to what they perceive as important. If a feature can fail gracefully or be [reasonably] polyfilled, I usually spring for it.

My knee jerk reaction to this is "can we reliably support both the Popover API _and_ hoist for those who need it?" and I _think_ the answer is _probably_.

Let's put this into the "want to do" category since we can make this change any time (even in beta/stable) without causing any problems. It would render `hoist` useless, but positioning would still be correct.

**Comment by LeaVerou** [comment]
We could ditch `hoist` and hoist by default, using the Popover API when it is supported and falling back to the current workaround when it isn't. We could even detect when an ancestor has one of these `overflow` values that need it and only do it then.

**Comment by claviska** [comment]
FYI — from my discussions with the Floating UI dev, the ancestor checking thing isn’t as straight forward as it sounds, so it’s a path I’d prefer not to stray down.

**Comment by claviska** [comment]
We are at 90% browser support for Popover. I'd like to get this in sooner than later, but I think we'll need to maintain the fallback as you suggest instead of just cutting over.

I like the idea of falling back to `hoist`, so I'll explore that a bit.

![CleanShot 2025-01-13 at 12 25 36@2x](https://github.com/user-attachments/assets/e662c258-4120-430c-ae50-8a40570094a8)

**Comment by LeaVerou** [comment]
I think we should remove the user-facing `hoist` but fall back to that *behavior* when Popover is not available. Sure, it's slower, but also, it will only happen for < 10% of users, so I think the tradeoff of users not having to learn another attribute to make things work as expected is worth it.

To add more context about why I think `hoist` should be removed (and while most of my opinions are very weakly held, that's one I feel more confident about than my average — though it's entirely possible I'm missing some essential bit of info about how `hoist` is used):

The main reason is that It has **no user-facing purpose**. There is no case where you _don't_ want `hoist`, there are only cases where you can get away without it, and cases where not having it actually creates a problem. So making that behavior opt-in introduces several footguns for users:
- Learnability: users need to learn an extra thing about WA just to make things work, but things should work out of the box (ideally).
- It introduces the potential for bugs, as even for users that _do_ know about `hoist`, they may not always foresee that they need it. E.g. see our demos that were broken because we didn't have it.
 
Ideally, opt-in behaviors should be things that fundamentally are sometimes desirable and sometimes not. Here `hoist` was not made opt-in for any user-facing reason, but for implementation reasons: because the way it was implemented it would be too slow to always hoist (and possibly buggy, since the workaround was not perfect). However, now that we can do this properly for 90% of users, the current implementation of `hoist` becomes just a polyfill for those on older browsers, so I think the perf hit is acceptable.

**Comment by claviska** [comment]
> I think we should remove the user-facing hoist but fall back to that behavior when Popover is not available.

Yep, that's what I meant. Falling back to the hoist _behavior_. I'm OK to remove the public facing API for it :)

**Comment by LeaVerou** [comment]
Ah, we are in agreement then!
